[lld][ELF] Implement merged .debug_names section

MaskRay opened the pull request.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Update LLD to be able to generate a single merged .debug_names section (rather
than a set of appended sections). Option is controlled by a new flag, --debug-names.
Does not handle type units yet (coming soon).

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Fix two typos in test cases.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Clean up some error checking and reporting, and add some new tests.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Update and fix debug-names* tests. Make entire DebugNamesSection
a template <class ELFT> section. Fix formatting and other
review comments.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Fix code formatting issue.

cmtice committed:
[lld][ELF] Implement merged .debug_names section

Update for-loop to make sure unnecessary evaluations/calculations are
not performed on condition checks.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Move '-r' and '--debug-names' test to driver.test; remove
debug-names-invalid-flags.s test.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Address various review comments.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Fix comment,

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Replace "ELFT::TargetEndianness" with "ELFT::Endianness"

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Set augmentation string size from the augmentation string (for error
handling case).

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Update debug-names*.s tests, to use split file format rather than
multiple files.

Reduce the use of 'auto' in DebugNamesSection work, replacing it with
actual type names.  Use UINT*_MAXINT rather than constant values.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Update tests to up split-file instead of multiple files.

MaskRay committed:
Simplify and optimize .debug_names and removed dead TU code

[ELF] Add --debug-names to create merged .debug_names

`clang -g -gpubnames` (with optional -gsplit-dwarf) creates the
`.debug_names` section ("per-CU" index). By default lld concatenates
input `.debug_names` sections into an output `.debug_names` section.
LLDB can consume the concatenated section but the lookup performance is
not good.

This patch adds --debug-names to create a per-module index by combining
the per-CU indexes into a single index that covers the entire load
module. The produced `.debug_names` is a replacement for `.gdb_index`.

Type units (-fdebug-types-section) are not handled yet.

Co-authored-by: Fangrui Song <i@maskray.me>

cmtice commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -2711,6 +2714,579 @@ static uint32_t computeGdbHash(StringRef s) {
+    for (NameTableEntry &nte : ni.nameEntries)
+      for (IndexEntry &ie : nte.entries())
+        ie.parentEntry =
+            ie.parentOffset == 0 ? nullptr : offsetMap.lookup(ie.parentOffset);
```
This is incorrect. 0 is a valid offset. You must use something like -1 to indicate a non-valid or unset offset.

MaskRay replied:
`parentOffset = 0` is an acceptable default value since valid values are >= EntriesBase, larger than the header size.

cmtice commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -2711,6 +2714,581 @@ static uint32_t computeGdbHash(StringRef s) {
+  FoldingSet<Abbrev> abbrevSet;
+  const dwarf::Form cuAttrForm = getMergedCuCountForm(hdr.CompUnitCount).second;
+  for (InputChunk &chunk : inputChunks) {
+    for (auto [i, llvmNi] : enumerate(*chunk.llvmDebugNames)) {
```
Is there any way to make this a reference, rather than a  copy?

MaskRay replied:
The second element of the tuple, `llvmNi`, is a reference.

cmtice commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -2711,6 +2714,581 @@ static uint32_t computeGdbHash(StringRef s) {
+          if (inserted)
+            nameVec.push_back(std::move(nte));
+          else
+            nameVec[it->second].indexEntries.append(nte.indexEntries);
```
This should be a std::move, to make sure we don't accidentally copy/change any of the IndexEntry addresses.

MaskRay replied:
The code is fine. 

There is no rvalue reference overload for `append` and such a overload would not make a difference as we would have to copy the elements one by one.

cmtice commented on lld/ELF/SyntheticSections.h:
```diff
@@ -789,6 +793,121 @@ class RelroPaddingSection final : public SyntheticSection {
+                        const llvm::DWARFDebugNames::Header &hdr,
+                        const llvm::DWARFDebugNames::DWARFDebugNamesOffsets &)>
+                        readOffsets);
+  void computeHdrAndAbbrevTable(MutableArrayRef<InputChunk>);
```
Why is the parameter not a reference?

MaskRay replied:
MutableArrayRef is a cheap type (2 words). Passing-by-reference is actually more expensive than passing-by-value here.

cmtice commented on lld/ELF/SyntheticSections.h:
```diff
@@ -789,6 +793,121 @@ class RelroPaddingSection final : public SyntheticSection {
+                        const llvm::DWARFDebugNames::DWARFDebugNamesOffsets &)>
+                        readOffsets);
+  void computeHdrAndAbbrevTable(MutableArrayRef<InputChunk>);
+  std::pair<uint32_t, uint32_t> computeEntryPool(MutableArrayRef<InputChunk>);
```
Same question: Why is the parameter not a reference?

MaskRay replied:
ditto above.  cheap type (2 words)

cmtice commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -2711,6 +2714,581 @@ static uint32_t computeGdbHash(StringRef s) {
+      return;
+    }
+
+    SmallVector<uint32_t, 0> entryOffsets = readOffsets(ni.hdr, locs);
```
Redundante declaration; you already declared entryOffsets on line 2746.

MaskRay replied:
Good catch. I removed the unused one on 2746.

cmtice committed:
[lld][ELF] Implement merged .debug_names section.

Migrate updates & fixes from maskray's lld-debug-names branch.

cmtice committed:
[lld][ELF] Properly handle multiple (appended) indices within single .o

Current implementation does not properly handle reading multiple
(appended) .debug_names sections withih the same input .o file. This updates
the code that reads from the .debug_nemes section to add in the proper
offset when there are multiple appended .debug_names (could be generated
by LTO, e.g.).

cmtice committed:
[lld][ELF] Add & update tests:

New tests:
   - ppc-debug-names.s : Tests .debug_names merging for big endiamn & 32-bit
   - debug-names-missing-parent.s: Tests .debug_names merging works when a
      DW_IDX_parent attribute is missing.
Updated debug-names-bad-die-idx-sizes.s & debug-names-bad-name-count.s:
   - Added tests with --noinhibit_execc to generate warnings & not die

Fixed names of source files in debug_names.s & the new/updated tests.

cmtice committed:
[lld][ELF] Fix clang-format issues.

Fix clang-format issues.

cmtice committed:
[lld][ELF] Clean up file & directory names in tests.

Shorten file names to a.cpp, b.cpp, etc.
Replace debug-names-test directory with /proc/self/cwd.

cmtice committed:
[lld][ELF] Add test for CompUnit > 1

Add code to update CU offsets correctly when there are multiple CUs
within one .o file. Also add test for this.

cmtice committed:
[lld][ELF] Clean up updateMultiCuOffset.

Use the existing .debug_info section pointer in chunk, rather than
rebuilding it.

MaskRay committed:
writeTo: Use endian::writeNext and rename oldBuf

cmtice committed:
Merge remote-tracking branch 'origin/main' into new-lld-work

MaskRay committed:
Merge branch 'main' into lld-debug-names2

MaskRay committed:
Simplify updateMultiCuOffsets

MaskRay committed:
Fix CU list offsets when there are 3+ CUs in one name index

Untested yet. https://github.com/llvm/llvm-project/pull/86508#issuecomment-2057815004
contains instructions to test this scenario.

cmtice committed:
Merge remote-tracking branch 'maskray/lld-debug-names' into new-lld-work

Bringing over maskray's latest contributions.

cmtice committed:
[lld][ELF] Minor fixes.

Fix some error strings & adjust tests that test for them.
Fix comments at top of ppc32-debug-names.s test.
Fix for-loop syntax and call NameIndex::getOffsets.

MaskRay committed:
Fix when there are more than one name indices in an input .debug_names

MaskRay committed:
Test compressed input and output

cmtice committed:
[lld][ELF] Add more tests for debug_names.

Add multiple variations of test for having multiple CUs in a single
.o file.
Add tests for using --debug_names when one or all of the input files
do not have a .debug_names section.

MaskRay committed:
Deduplicate tests

Add Inputs/debug-names-[ab].s which can be easily regenerated
Update some negative tests to use Inputs/debug-names-[ab].s or apply straightforward transformation with sed

clayborg commented:
Does this merge all .debug_names sections from all .o files into a single .debug_names section? If so this will cause problems with foreign type unit support when we use .debug_names with split DWARF. When we create a .dwp file, one of the type units from one of the .dwo files will end up in the .dwp file and then we have all of the accelerator tables from all copies of the same type unit in the final output file. Since type units are not always emitted the same, some accelerator table entries will have bad DIE offsets. Right now the only way for us to tell if a .debug_names entry matches the type unit it to look in the individual .debug_names table and see if it only has one compile unit. The type unit has a DW_AT_dwo_name attribute in it and if it matches the .dwo name for the one and only CU on the .debug_names table, we can trust it, else we ignore the entry. So if this patch makes a single .debug_names table, we lose this information and our .debug_names might have things pointing at the wrong name or contain DIE offsets that are in between two actual DIE entries in the foreign type units.

clayborg commented:
To see this in action, see this patch that is up for review right now: https://github.com/llvm/llvm-project/pull/87740

clayborg commented:
I added some reviewers that have more context.

ayermolo commented:
> I added some reviewers that have more context.

@clayborg the review is happening here: https://github.com/llvm/llvm-project/pull/86508

cmtice committed:
[lld][ELF] Minor fixes addressing review comments.

- Update comment about CompUnits in OutputChunk.
- Replace two uses of 'seq' in for-loops.
- Replace use of memcpy with llvm::copy.

cmtice committed:
[lld][ELF] Fixes to previous fixes.

Fix typo, add 'const', replace one use of 'seq'/

cmtice committed:
[lld][ELF] Update code related to type units.

- Keep type unit counts in output header at zero, regardless of input, to
avoid crashing LLD, since we're not actually handling type units yet.
- Move the type units warning to the point where we can check the input
counts.
- Add a type units test (that LLD doesn't crash and outputs the warning).

MaskRay committed:
Clean up more tests

Remove tools/generate-content.py.
We will use a generic tool intended for llvm/utils: https://discourse.llvm.org/t/utility-to-generate-elaborated-assembly-ir-tests/78408

MaskRay committed:
Clean up more tests

Remove tools/generate-content.py.
We will use a generic tool intended for llvm/utils: https://discourse.llvm.org/t/utility-to-generate-elaborated-assembly-ir-tests/78408

cmtice committed:
Merge remote-tracking branch 'maskray/lld-debug-names' into new-lld-work

cmtice committed:
[lld][ELF] Address last review comments.

- Pulled body of loop that reads/parses IndexEntries into a
  separate function (readEntry)
- Removed unnecessary 'const' qualifiers on local vars
- Updated comment about handling type units
- Created getChunks accessor function; used it to eliminate use of
  'seq' in for-loop.
- Enhanced some error messages
- Update debug-names-bad.s test to handle udpated error messages

cmtice committed:
[lld][ELF] Clean up error messages.

- Clean up the error messages in  readEntry function
- Restore 'const' qualifier to size_t vars in computeEntryPool
- Add 'const' qualifier to getChunks function.

MaskRay committed:
Update some diagnostics in readEntry and clang-format

MaskRay commented:
Everything has been merged into #86508 , which landed yesterday.
